Implemented retry logic + Custom response handler + Partner Center Billing reports API

shobhitvaish opened the pull request:
@Jabe Please note that some Partner Center billing reports are available via Graph API

PR implements the following:
1. Ability to retry the request. [Example](https://github.com/realmjoin/SlimLib.Microsoft.Graph/compare/main...shobhitvaish:SlimLib.Microsoft.Graph:dev?expand=1#diff-a189887fe73cbdcf2d0d02251c4560711570ab9b36eb192cfe243d804125338dR39)
2. Ability to use a custom response handler. This is useful to read the headers. [Example](https://github.com/realmjoin/SlimLib.Microsoft.Graph/compare/main...shobhitvaish:SlimLib.Microsoft.Graph:dev?expand=1#diff-a189887fe73cbdcf2d0d02251c4560711570ab9b36eb192cfe243d804125338dR23)
3. Partner Center Billing reports API endpoints

1 and 2 above are needed because billing reports API uses an asynchronous request-reply pattern. More information: Partner Center Billing reports API

shobhitvaish committed:
Implemented retry logic + Custom response handler + Partner Center Billing Reports API

Jabe commented:
Hey Shobhit, thank you for the change.

Regarding 1): I'd like to keep retry logic out of this project. Have a look at the internal projects where retries are implemented using Polly. I have used it successfully with SlimLib like this as a global policy for example:

```c#
private static AsyncRetryPolicy<HttpResponseMessage> GetRetryPolicy(ILoggerFactory loggerFactory)
{
    const int MaxRetries = 6;

    var logger = loggerFactory.CreateLogger("HTTP");

    static TimeSpan SleepDurationProvider(int retryAttempt, DelegateResult<HttpResponseMessage> result, Context context)
    {
        var serverWait = result.Result?.Headers.RetryAfter?.Delta ?? TimeSpan.Zero;
        var clientWait = Backoff.Exponential(retryAttempt);

        return serverWait > clientWait ? serverWait : clientWait;
    }

    Task OnRetryAsync(DelegateResult<HttpResponseMessage> result, TimeSpan elapsed, int retryAttempt, Context context)
    {
        var uri = result.Result.RequestMessage?.RequestUri;

        if (retryAttempt == MaxRetries)
            logger.LogError("Retry attempt {RetryAttempt} of {MaxRetries} failed after {ElapsedDuration} sec, with {StatusCode}. Graph target: {RequestUri}", retryAttempt, MaxRetries, elapsed.TotalSeconds, result.Result?.StatusCode, uri);
        else
            logger.LogWarning("Retry attempt {RetryAttempt} of {MaxRetries} failed after {ElapsedDuration} sec, with {StatusCode}. Retrying in {SleepDuration} sec. Graph target: {RequestUri}", retryAttempt, MaxRetries, elapsed.TotalSeconds, result.Result?.StatusCode, elapsed.TotalSeconds, uri);

        return Task.CompletedTask;
    }

    return HttpPolicyExtensions
        .HandleTransientHttpError()
        .OrResult(x => x.StatusCode == System.Net.HttpStatusCode.TooManyRequests)
        .WaitAndRetryAsync(MaxRetries, SleepDurationProvider, OnRetryAsync);
}
```

Now, looking at this code, this is kind of an "expected" retry scenario that is part of the process. It might benefit from a dedicated Polly policy in the surrounding code, but with this existing config ⬆️ it might just already work.

Do you think you move the retry logic to the target project using a Polly policy and keep `GetOperationAsync` simple? I think it should just fail.

shobhitvaish committed:
Remove request retry options

shobhitvaish commented:
@Jabe, Thanks for the feedback. I didn't notice we are using Polly, which is, of course, a go-to library for retries.
1. I have removed the ability to retry the request.
2. The existing Polly policy won't work because reading the response body is also needed to get the operation status. Hence, I added a new policy, but it is not working as expected. I'll write about it the RJ repo.

Jabe commented:
Thank you!!
